doc/python-type-hinting.md

```diff
@@ -0,0 +1,85 @@
+# Python Type Hinting
+
+As a team we have committed to adding type hints throughout the Python code. This document defines our approach to type hinting, to ensure consistency in the codebase and avoid repeating discussions about how we apply typing.
+
+## Variables
+
+Specify types for variables initialised with `None`:
+
+```python
+description: Optional[str] = None
+```
+
+Specify types for empty collections:
+
+```python
+items: list[str] = []
+mappings: dict[str, int] = {}
+```
+
+In all other places, types for variables are optional.
+
+## Standard collections
+
+For standard collections, use lower case names e.g `list` rather than `List`:
+
+```python
+items: list[int] = [1]
+mappings: set[int] = {1, 2}
+```
+
+https://www.python.org/dev/peps/pep-0585/
+
+## Generic types
+
+Specify at least the top-level type parameters for all generic types:
+
+```python
+def get_objects_matching(ids: Sequence[str]) -> dict[int, dict]
+```
+
+## Optional 
+
+When something is a single type and is optional use `Optional`:
+
+```python
+def test(self, var: Optional[int]) -> None:
+```
+
+When something can be multiple types and is optional use `Union` with `None`:
+
+```python
+def test(self, var: Union[None, int, str]) -> None:
```

Comment:
Any specific reason for this over `Optional[Union[int,str]]` since `Optional[int]` is just syntatic sugar for `Union[None, int]`?.

Reply:
Personal preference I guess, I think it reads more clearly without the nested `[]`.

Reply:
Fair enough. I do think Union without Optional is better syntactically, but I do think "Optional" makes the easier to understand. Also, IDE like PyCharm suggests converting Union with None to Optional Union. 

The recommendation in the image is slightly wrong as the Union shouldn't include `None` but it does execute it successfully to `Optional[Union[str, int]]`. The inspection rule:
```
Inspection info: This inspection reports when the Union[None, ...] annotation was used instead Optional[Union[...]]. 
Union[object, None] -> Optional[object] 
Union[int, string, None] -> Optional[Union[int, string]] 
Union[None] -> None
```
![image](https://user-images.githubusercontent.com/35296336/117165676-9975f200-adbd-11eb-92a6-80f64cc5bccc.png)

Happy to keep as is, if that is the preference by the majority.

Reply:
I prefer the flat structure of `Union[None, int, str]`